homeassistant/components/homeworks/__init__.py

```diff
@@ -66,64 +75,137 @@
         dispatcher_send(hass, signal, msg_type, values)
 
-    config = base_config[DOMAIN]
-    controller = Homeworks(config[CONF_HOST], config[CONF_PORT], hw_callback)
-    hass.data[HOMEWORKS_CONTROLLER] = controller
+    config = entry.options
+    try:
+        controller = await hass.async_add_executor_job(
+            Homeworks, config[CONF_HOST], config[CONF_PORT], hw_callback
+        )
+    except (ConnectionError, OSError) as err:
+        raise ConfigEntryNotReady from err
 
     def cleanup(event):
         controller.close()
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, cleanup)
+    entry.async_on_unload(hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, cleanup))
 
-    dimmers = config[CONF_DIMMERS]
-    load_platform(hass, Platform.LIGHT, DOMAIN, {CONF_DIMMERS: dimmers}, base_config)
-
-    for key_config in config[CONF_KEYPADS]:
+    keypads: dict[str, HomeworksKeypad] = {}
+    for key_config in config.get(CONF_KEYPADS, []):
         addr = key_config[CONF_ADDR]
         name = key_config[CONF_NAME]
-        HomeworksKeypadEvent(hass, addr, name)
+        keypads[addr] = HomeworksKeypad(hass, controller, controller_id, addr, name)
+
+    hass.data[DOMAIN][entry.entry_id] = HomeworksData(
+        controller, controller_id, keypads
+    )
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    entry.async_on_unload(entry.add_update_listener(update_listener))
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    if not await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        return False
+
+    data: HomeworksData = hass.data[DOMAIN].pop(entry.entry_id)
+    for keypad in data.keypads.values():
+        keypad.unsubscribe()
+
+    await hass.async_add_executor_job(data.controller.close)
 
     return True
 
 
-class HomeworksDevice(Entity):
+async def update_listener(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Handle options update."""
+    await hass.config_entries.async_reload(entry.entry_id)
+
+
+def calculate_unique_id(controller_id, addr, idx):
+    """Calculate entity unique id."""
+    return f"homeworks.{controller_id}.{addr}.{idx}"
+
+
+class HomeworksEntity(Entity):
     """Base class of a Homeworks device."""
 
     _attr_should_poll = False
 
-    def __init__(self, controller, addr, name):
+    def __init__(
+        self,
+        controller: Homeworks,
+        controller_id: str,
+        addr: str,
+        idx: int,
+        name: str | None,
+    ) -> None:
         """Initialize Homeworks device."""
         self._addr = addr
+        self._idx = idx
+        self._controller_id = controller_id
         self._attr_name = name
-        self._attr_unique_id = f"homeworks.{self._addr}"
+        self._attr_unique_id = calculate_unique_id(
+            self._controller_id, self._addr, self._idx
+        )
```

Comment:
Does this change the unique_id for existing devices? Else we might need to migrate the existing entities?

Reply:
Yes, you're right. The import flow migrates existing light entities: https://github.com/home-assistant/core/blob/5527abad30eb67395af23c09bec16d2e9028cf35/homeassistant/components/homeworks/config_flow.py#L417C1-L427

Reply:
Nice!